c/rc_lock.c

```diff
@@ -0,0 +1,462 @@
+  memset((void *)lock_bytes_seg.ptr, 0, lock_bytes_seg.size);
+  blake2b_update(&blake2b_ctx, (char *)&witness_len, sizeof(uint64_t));
+  blake2b_update(&blake2b_ctx, temp, read_len);
+
+  // remaining of first witness
+  if (read_len < witness_len) {
+    ret = load_and_hash_witness(&blake2b_ctx, read_len, 0,
+                                CKB_SOURCE_GROUP_INPUT, false);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+  }
+
+  // Digest same group witnesses
+  size_t i = 1;
+  while (1) {
+    ret =
+        load_and_hash_witness(&blake2b_ctx, 0, i, CKB_SOURCE_GROUP_INPUT, true);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+    i += 1;
+  }
+
+  // Digest witnesses that not covered by inputs
+  i = ckb_calculate_inputs_len();
+  while (1) {
+    ret = load_and_hash_witness(&blake2b_ctx, 0, i, CKB_SOURCE_INPUT, true);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+    i += 1;
+  }
+
+  blake2b_final(&blake2b_ctx, message, BLAKE2B_BLOCK_SIZE);
+
+  /* Load signature */
+  secp256k1_context context;
+  uint8_t secp_data[CKB_SECP256K1_DATA_SIZE];
+  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
+  if (ret != 0) {
+    return ret;
+  }
+
+  secp256k1_ecdsa_recoverable_signature signature;
+  if (secp256k1_ecdsa_recoverable_signature_parse_compact(
+          &context, &signature, signature_bytes,
+          signature_bytes[RECID_INDEX]) == 0) {
+    return ERROR_SECP_PARSE_SIGNATURE;
+  }
+
+  /* Recover pubkey */
+  secp256k1_pubkey pubkey;
+  if (secp256k1_ecdsa_recover(&context, &pubkey, &signature, message) != 1) {
+    return ERROR_SECP_RECOVER_PUBKEY;
+  }
+
+  /* Check pubkey hash */
+  size_t pubkey_size = PUBKEY_SIZE;
+  if (secp256k1_ec_pubkey_serialize(&context, temp, &pubkey_size, &pubkey,
+                                    SECP256K1_EC_COMPRESSED) != 1) {
+    return ERROR_SECP_SERIALIZE_PUBKEY;
+  }
+
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, temp, pubkey_size);
+  blake2b_final(&blake2b_ctx, temp, BLAKE2B_BLOCK_SIZE);
+
+  if (memcmp(pubkey_hash, temp, BLAKE160_SIZE) != 0) {
+    return ERROR_PUBKEY_BLAKE160_HASH;
+  }
+
+  return 0;
+}
+
+static uint32_t read_from_witness(uintptr_t arg[], uint8_t *ptr, uint32_t len,
+                                  uint32_t offset) {
+  int err;
+  uint64_t output_len = len;
+  err = ckb_load_witness(ptr, &output_len, offset, arg[0], arg[1]);
+  if (err != 0) {
+    return 0;
+  }
+  if (output_len > len) {
+    return len;
+  } else {
+    return output_len;
+  }
+}
+
+uint8_t g_witness_data_source[DEFAULT_DATA_SOURCE_LENGTH];
+int make_witness(WitnessArgsType *witness) {
+  int err = 0;
+  uint64_t witness_len = 0;
+  size_t source = CKB_SOURCE_GROUP_INPUT;
+  err = ckb_load_witness(NULL, &witness_len, 0, 0, source);
```

Comment:
I wasn't sure on this: if we use molecule-c2 here, why do we load the full witness at once? I feel like if we use molecule-c2, we should do streaming load, otherwise we can use the original molecule and load the full witness at once.

Reply:
Here it just loads the length of witness, note: first arguments is NULL .
According to the RFC: https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md
```
One trick here, is that by providing NULL as addr, and a uint64_t pointer with 0 value as len, this syscall can be used to fetch the length of the serialized data part without reading any actual data.
```

After that, moleculec-c2 uses this length to make sure no out-of-bounds reading will occur.